journalpump/senders/websocket.py

```diff
@@ -0,0 +1,286 @@
+            self.log.error("Websocket certificate verification error: %r. Retrying.", ex)
+        except OSError as ex:  # Network unreachable, etc, may happen sporadically
+            self.log.warning("Websocket connection error: %r. Retrying.", ex)
+        except Exception as ex:  # pylint:disable=broad-except
+            self.log.exception("Unhandled exception occurred on websocket connection: %r", ex)
+        finally:
+            websocket = None
+            self.connected = False
+
+        for task in [consumer_task, producer_task, stop_event_task]:
+            try:
+                if task:
+                    task.cancel()
+            except Exception as ex:  # pylint:disable=broad-except
+                self.log.exception("Unhandled exception occurred: %r", ex)
+
+    async def sleep_before_reconnect(self):
+        if self.running:
+            sleep_interval = self.backoff.next_sleep()
+            self.log.info("Retrying websocket connection in %.1f", sleep_interval)
+            while sleep_interval > 0 and self.running:
+                await asyncio.sleep(min(sleep_interval, 1))
+                sleep_interval -= 1
+
+    async def comms_channel_loop(self):
+        self.queue = Queue(maxsize=1000)
+
+        while self.running:
+            await self.comms_channel_round()
+            await self.sleep_before_reconnect()
+
+        self.log.info("Websocket closed")
+
+
+class WebsocketSender(LogSender):
+    def __init__(self, *, config, **kwargs):
+        super().__init__(config=config, max_send_interval=config.get("max_send_interval", 0.5), **kwargs)
+        self.runner = None
+        self.config = config
+
+    def _init_websocket(self) -> None:
+        self.log.info("Initializing Websocket client, address: %r", self.config["websocket_uri"])
+
+        if self.runner:
+            self.runner.close()
+            self.runner = None
+
+        self.mark_disconnected()
+
+        while self.running and not self._connected:
+            # retry connection
+            time.sleep(1)
+
+            try:
+                runner = WebsocketRunner(
+                    websocket_uri=self.config["websocket_uri"],
+                    socks5_proxy_url=self.config.get("socks5_proxy_url"),
+                    ssl_enabled=self.config.get("ssl"),
+                    ssl_ca=self.config.get("ca"),
+                    ssl_key=self.config.get("keyfile"),
+                    ssl_cert=self.config.get("certfile"),
+                )
+                runner.start()
+            except Exception as ex:  # pylint:disable=broad-except
+                self.mark_disconnected(ex)
+                self.log.exception("Retriable error during Websocket initialization: %s: %s", ex.__class__.__name__, ex)
+                self._backoff()
+            else:
+                self.log.info("Initialized Websocket client, address: %r", self.config["websocket_uri"])
+                self.runner = runner
+                self.mark_connected()
```

Comment:
It would be nicer to report the actual connected status. Now this just reports whether the thread has been created which in practice always succeeds and is always running yet connection might well not exist.

Reply:
Yes, absolutely. Placeholder to get things running.